languages/python/sqlalchemy-oso/tests/test_roles2.py

```diff
@@ -0,0 +1,1465 @@
+    # - Assigning/removing non-existent role throws an error
+    oso, oso_roles, session = init_oso
+    policy = """
+    resource(_type: Organization, "org", actions, roles) if
+        actions = ["invite"] and
+        roles = {
+            org_member: {
+                perms: ["invite"]
+            }
+        };
+
+    allow(actor, action, resource) if
+        Roles.role_allows(actor, action, resource);
+    """
+    oso.load_str(policy)
+    oso_roles.synchronize_data()
+
+    osohq = sample_data["osohq"]
+    leina = sample_data["leina"]
+
+    with pytest.raises(OsoError):
+        oso_roles.assign_role(leina, osohq, "org_owner", session=session)
+
+    with pytest.raises(OsoError):
+        oso_roles.remove_role(leina, osohq, "org_owner", session=session)
+
+
+def test_remove_unassigned_role(init_oso, sample_data):
+    # - Removing role that user doesn't have returns false
+    oso, oso_roles, session = init_oso
+    policy = """
+    resource(_type: Organization, "org", actions, roles) if
+        actions = ["invite"] and
+        roles = {
+            org_member: {
+                perms: ["invite"]
+            }
+        };
+
+    allow(actor, action, resource) if
+        Roles.role_allows(actor, action, resource);
+    """
+    oso.load_str(policy)
+    oso_roles.synchronize_data()
+
+    osohq = sample_data["osohq"]
+    leina = sample_data["leina"]
+
+    removed = oso_roles.remove_role(leina, osohq, "org_member", session=session)
+    assert not removed
+
+
+def test_assign_remove_user_role(init_oso, sample_data):
+    # - Adding user-role assignment grants access
+    oso, oso_roles, session = init_oso
+    policy = """
+    resource(_type: Organization, "org", actions, roles) if
+        actions = ["invite", "list_repos"] and
+        roles = {
+            org_member: {
+                perms: ["invite"]
+            },
+            org_owner: {
+                perms: ["list_repos"]
+            }
+        };
+
+    allow(actor, action, resource) if
+        Roles.role_allows(actor, action, resource);
+    """
+    oso.load_str(policy)
+    oso_roles.synchronize_data()
+
+    osohq = sample_data["osohq"]
+    leina = sample_data["leina"]
+    steve = sample_data["steve"]
+
+    oso_roles.assign_role(leina, osohq, "org_member", session=session)
+
+    # Assign leina member role
+    leina_roles = (
+        session.query(oso_roles.UserRole)
+        .filter(oso_roles.UserRole.user_id == leina.id)
+        .all()
+    )
+    assert len(leina_roles) == 1
+    assert leina_roles[0].role == "org_member"
+
+    # Assign steve owner role
+    oso_roles.assign_role(steve, osohq, "org_owner", session=session)
+
+    steve_roles = (
+        session.query(oso_roles.UserRole)
+        .filter(oso_roles.UserRole.user_id == steve.id)
+        .all()
+    )
+    assert len(steve_roles) == 1
+    assert steve_roles[0].role == "org_owner"
+
+    assert oso.is_allowed(leina, "invite", osohq)
```

Comment:
I'm a little confused what this works without a commit.

Reply:
hmm, yeah

Reply:
Hi it's steve from the future. Turns out none of these worked without a commit except on sqlite in memory. Makes sense.

Reply:
thanks future steve